substrate/frame/contracts/src/exec.rs

```diff
@@ -988,24 +993,27 @@ where
 						Event::Instantiated { deployer: caller, contract: account_id.clone() },
 					);
 				},
-				(ExportedFunction::Call, Some(code_hash)) => {
-					Contracts::<T>::deposit_event(
-						vec![T::Hashing::hash_of(account_id), T::Hashing::hash_of(&code_hash)],
-						Event::DelegateCalled { contract: account_id.clone(), code_hash },
-					);
-				},
+				(ExportedFunction::Call, Some(code_hash)) =>
+					if !self.is_read_only() {
+						Contracts::<T>::deposit_event(
+							vec![T::Hashing::hash_of(account_id), T::Hashing::hash_of(&code_hash)],
+							Event::DelegateCalled { contract: account_id.clone(), code_hash },
+						);
+					},
 				(ExportedFunction::Call, None) => {
 					// If a special limit was set for the sub-call, we enforce it here.
 					// The sub-call will be rolled back in case the limit is exhausted.
 					let frame = self.top_frame_mut();
 					let contract = frame.contract_info.as_contract();
 					frame.nested_storage.enforce_subcall_limit(contract)?;
 
-					let caller = self.caller();
-					Contracts::<T>::deposit_event(
-						vec![T::Hashing::hash_of(&caller), T::Hashing::hash_of(&account_id)],
-						Event::Called { caller: caller.clone(), contract: account_id.clone() },
-					);
+					if !self.is_read_only() {
+						let caller = self.caller();
+						Contracts::<T>::deposit_event(
+							vec![T::Hashing::hash_of(&caller), T::Hashing::hash_of(&account_id)],
+							Event::Called { caller: caller.clone(), contract: account_id.clone() },
+						);
+					}
```

Comment:
Let's create a new event for static call instead of suppressing the event here. It's only custom events by the contract that we want to suppress.